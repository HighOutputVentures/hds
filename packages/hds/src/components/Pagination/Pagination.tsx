import {
  ButtonProps,
  Flex,
  FlexProps,
  HStack,
  IconButton,
  IconProps,
  Select,
  SelectProps,
  StackProps,
  Text,
  TextProps,
  ThemingProps,
  useMultiStyleConfig,
} from '@chakra-ui/react';
import React, { FC, useCallback, useId } from 'react';
import { ChevronLeftIcon, ChevronRightIcon } from '../../../../icons/src/index';

type WithoutChildren<T> = Omit<T, 'children'>;

export interface PaginationProps extends ThemingProps {
  page: number;
  size: number;
  total: number;
  onSizeChange?: (newSize: number) => void;
  onPageChange?: (newPage: number) => void;
  options: {
    sizes: number[];
  };
  /**
   *
   * _Not yet implemented_
   *
   */
  loading?: boolean;
  partProps?: Partial<{
    container?: WithoutChildren<FlexProps>;
    dropdown?: WithoutChildren<SelectProps>;
    dropdownLabel?: WithoutChildren<TextProps>;
    dropdownContainer?: WithoutChildren<StackProps>;
    caption?: WithoutChildren<TextProps>;
    captionAndControlsContainer?: WithoutChildren<StackProps>;
    controls?: WithoutChildren<ButtonProps>;
    controlIcons?: WithoutChildren<IconProps>;
    controlsContainer?: WithoutChildren<StackProps>;
  }>;
}

const Pagination: FC<PaginationProps> = ({
  page,
  size,
  total,
  onPageChange,
  onSizeChange,
  options,
  partProps,
  variant,
}) => {
  const styles = useMultiStyleConfig('Pagination', { variant });
  const id = useId();

  const hasPrev = page > 1;
  const hasNext = page * size < total;

  const handleSizeChange = (e: React.ChangeEvent<HTMLSelectElement>) => {
    if (onSizeChange) onSizeChange(parseInt(e.target.value));
  };

  const handlePageChange = (type: 'increment' | 'decrement') => {
    const newPage = type === 'increment' ? page + 1 : page - 1;

    return () => (onPageChange ? onPageChange(newPage) : newPage);
  };

  const indexStart = (page - 1) * size + 1;
  const indexStop = page * size;
  const remainder = total % size;

  const getPageInfo = useCallback(() => {
    let pageInfo = '';

    pageInfo += 'Page ';
    pageInfo += indexStart;
    pageInfo += '-';
    pageInfo += indexStop > total ? indexStart - 1 + remainder : indexStop;
    pageInfo += ' of ';
    pageInfo += total;

    return pageInfo;
  }, [indexStart, indexStop, total, remainder]);

  return (
    <Flex
      id={id}
      alignItems="center"
      justifyContent="space-between"
      sx={styles.container}
      {...partProps?.container}
    >
      <HStack spacing={2} {...partProps?.dropdownContainer}>
        <Text
          as="span"
          whiteSpace="nowrap"
          sx={styles.dropdownLabel}
          {...partProps?.dropdownLabel}
        >
          Show rows per page
        </Text>

        <Select
          data-testid={`${id}-pagination.dropdown`}
          onChange={handleSizeChange}
          value={size}
          sx={styles.dropdown}
          {...partProps?.dropdown}
        >
          {options.sizes.map((size, index) => (
            <option
              value={size}
              key={id + size + index}
              data-testid={`${id}-${size}-${index}`}
            >
              {size}
            </option>
          ))}
        </Select>
      </HStack>

      <HStack spacing={4} {...partProps?.captionAndControlsContainer}>
        <Text as="span" sx={styles.caption} {...partProps?.caption}>
          {getPageInfo()}
        </Text>

        <HStack {...partProps?.controlsContainer}>
          <IconButton
            aria-label=""
            data-testid={`${id}-pagination.controls.prev`}
            icon={
              <ChevronLeftIcon
                color="white"
                sx={styles.controlIcons}
                {...partProps?.controlIcons}
              />
            }
            onClick={handlePageChange('decrement')}
            paddingTop="10px"
            paddingLeft="12px"
            disabled={!hasPrev}
            sx={styles.iconButton}
            {...partProps?.controls}
          />

          <IconButton
            aria-label=""
            data-testid={`${id}-pagination.controls.next`}
            icon={
              <ChevronRightIcon
                color="white"
                sx={styles.controlIcons}
                {...partProps?.controlIcons}
              />
            }
            onClick={handlePageChange('increment')}
            disabled={!hasNext}
            paddingTop="10px"
            paddingLeft="12px"
            sx={styles.iconButton}
            {...partProps?.controls}
          />
        </HStack>
      </HStack>
    </Flex>
  );
};

export default Pagination;
