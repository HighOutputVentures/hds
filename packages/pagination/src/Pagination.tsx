import { BoxProps } from "@chakra-ui/react";
import * as React from "react";
import GroupPagination, { GroupPaginationProps } from "./GroupPagination";
import MinimalPagination, { MinimalPaginationProps } from "./MinimalPagination";
import omit from "./utils";

export type PaginationProps =
  | ({ variant?: "minimal" } & MinimalPaginationProps)
  | ({ variant: "group" } & GroupPaginationProps);

export default function Pagination(props: PaginationProps & BoxProps) {
  switch (props.variant) {
    case "group":
      return <GroupPagination {...omit(props, "variant")} />;
    default:
      return <MinimalPagination {...omit(props, "variant")} />;
  }
}
