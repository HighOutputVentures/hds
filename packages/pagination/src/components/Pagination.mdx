import { Canvas, Meta, Story } from '@storybook/addon-docs';

<Meta title="Components/Pagination" />

# HDS Pagination

To install package, run command:

```bash
npm i @highoutput/hds-pagination
```

<Canvas>
  <Story id="components-pagination--default"></Story>
</Canvas>
