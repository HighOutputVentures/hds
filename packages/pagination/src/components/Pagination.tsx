import {
  Flex,
  HStack,
  IconButton,
  Select,
  Text,
  ThemingProps,
  useMultiStyleConfig,
} from '@chakra-ui/react';
import React, { FC, useCallback } from 'react';
import ChevronLeftIcon from './ChevronLeftIcon';
import ChevronRightIcon from './ChevronRightIcon';

export interface PaginationProps extends ThemingProps {
  page: number;
  size: number;
  total: number;
  onSizeChange?: (newSize: number) => void;
  onPageChange?: (newPage: number) => void;
  options: {
    sizes: number[];
  };
  loading?: boolean;
}

const Pagination: FC<PaginationProps> = ({
  page,
  size,
  total,
  onPageChange,
  onSizeChange,
  options,
  variant,
}) => {
  const styles = useMultiStyleConfig('Pagination', { size, variant });

  const hasPrev = page > 1;
  const hasNext = page * size < total;

  const handleSizeChange = (e: React.ChangeEvent<HTMLSelectElement>) => {
    if (onSizeChange) onSizeChange(parseInt(e.target.value));
  };

  const handlePageChange = (type: 'increment' | 'decrement') => {
    const newPage = type === 'increment' ? page + 1 : page - 1;

    return () => (onPageChange ? onPageChange(newPage) : newPage);
  };

  const indexStart = (page - 1) * size + 1;
  const indexStop = page * size;
  const remainder = total % size;

  const getPageInfo = useCallback(() => {
    let pageInfo = '';

    pageInfo += 'Page ';
    pageInfo += indexStart;
    pageInfo += '-';
    pageInfo += indexStop > total ? indexStart - 1 + remainder : indexStop;
    pageInfo += ' of ';
    pageInfo += total;

    return pageInfo;
  }, [indexStart, indexStop, total, remainder]);

  return (
    <Flex
      alignItems="center"
      justifyContent="space-between"
      sx={styles.container}
    >
      <HStack spacing={2}>
        <Text as="span" whiteSpace="nowrap" sx={styles.dropdownLabel}>
          Show rows per page
        </Text>

        <Select
          aria-label="Change page size"
          onChange={handleSizeChange}
          value={size}
          sx={styles.dropdown}
        >
          {options.sizes.map((size, index) => (
            <option key={`${size}${index}`} value={size}>
              {size}
            </option>
          ))}
        </Select>
      </HStack>

      <HStack spacing={4}>
        <Text as="span" sx={styles.caption}>
          {getPageInfo()}
        </Text>

        <HStack>
          <IconButton
            aria-label="Go to previous page"
            icon={<ChevronLeftIcon color="white" sx={styles.controlIcons} />}
            onClick={handlePageChange('decrement')}
            paddingTop="10px"
            paddingLeft="12px"
            disabled={!hasPrev}
            sx={styles.iconButton}
          />

          <IconButton
            aria-label="Go to next page"
            icon={<ChevronRightIcon color="white" sx={styles.controlIcons} />}
            onClick={handlePageChange('increment')}
            disabled={!hasNext}
            paddingTop="10px"
            paddingLeft="12px"
            sx={styles.iconButton}
          />
        </HStack>
      </HStack>
    </Flex>
  );
};

export default Pagination;
