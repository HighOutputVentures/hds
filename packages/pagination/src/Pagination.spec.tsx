import "@testing-library/jest-dom";
import { act, fireEvent, render } from "@testing-library/react";
import * as React from "react";
import renderer from "react-test-renderer";
import Pagination from "./Pagination";

// Heads up! we're only gonna test "minimal" variant
// since "group" variant's API is just a copy of "minimal"
// but "minimal" is a bit complex
describe("Pagination", () => {
  const renderPagination = (
    props: Partial<{
      page: number;
      pageSize: number;
      total: number;
      onChange: (value: { page: number; pageSize: number }) => void;
    }> = {},
  ) => {
    return render(
      <Pagination
        variant="minimal"
        page={1}
        pageSize={5}
        count={100}
        onChange={function () {}}
        {...props}
      />,
    );
  };

  it.skip("Should be able to go to next page", async () => {
    const onChange = jest.fn();
    const { getByRole } = renderPagination({ page: 1, onChange });

    await act(async () => {
      fireEvent.click(getByRole("button", { name: "Go to next page" }));
    });

    expect(onChange).toHaveBeenCalledWith(
      expect.objectContaining({
        page: 2,
        pageSize: expect.any(Number),
      }),
    );
  });

  it("Should be able to go to previous page", async () => {
    const onChange = jest.fn();
    const { getByRole } = renderPagination({ page: 2, onChange });

    await act(async () => {
      fireEvent.click(getByRole("button", { name: "Go to previous page" }));
    });

    expect(onChange).toHaveBeenCalledWith(
      expect.objectContaining({
        page: 1,
        pageSize: expect.any(Number),
      }),
    );
  });

  it.skip("Should match snapshot", () => {
    const component = renderer.create(
      <Pagination
        variant="minimal"
        page={1}
        pageSize={5}
        count={100}
        onChange={function () {
          //
        }}
      />,
    );

    const snapshot = component.toJSON();
    expect(snapshot).toMatchSnapshot();
    component.unmount();
  });
});
