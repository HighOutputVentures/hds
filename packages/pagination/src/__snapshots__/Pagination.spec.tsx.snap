// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination Should match snapshot 1`] = `
<div
  className="css-8omsyd"
>
  <div
    className="css-su60pv"
  >
    <div
      className="css-8dt55n"
    >
      <button
        aria-label="Go to previous page"
        className="chakra-button css-wklby5"
        data-disabled=""
        data-part="prev-page-trigger"
        data-scope="pagination"
        data-testid="hds.minimal-pagination.previous.button"
        id="pagination::r0::prev"
        onClick={[Function]}
        type="button"
      >
        <svg
          className="chakra-icon chakra-icon css-wuvk2m"
          focusable={false}
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M15.8334 9.99996H4.16675M4.16675 9.99996L10.0001 15.8333M4.16675 9.99996L10.0001 4.16663"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="1.66667"
          />
        </svg>
        Previous
      </button>
      <div
        className="chakra-stack css-15lgeks"
      >
        <button
          aria-current="page"
          aria-label="page 1"
          className="chakra-button css-1pwj5mw"
          data-part="page-trigger"
          data-scope="pagination"
          data-selected=""
          data-testid="hds.minimal-pagination.page.control"
          id="pagination::r0::item:1"
          onClick={[Function]}
          type="button"
        >
          1
        </button>
        <button
          aria-label="page 2"
          className="chakra-button css-zndnet"
          data-part="page-trigger"
          data-scope="pagination"
          data-testid="hds.minimal-pagination.page.control"
          id="pagination::r0::item:2"
          onClick={[Function]}
          type="button"
        >
          2
        </button>
        <button
          aria-label="page 3"
          className="chakra-button css-zndnet"
          data-part="page-trigger"
          data-scope="pagination"
          data-testid="hds.minimal-pagination.page.control"
          id="pagination::r0::item:3"
          onClick={[Function]}
          type="button"
        >
          3
        </button>
        <button
          aria-label="page 4"
          className="chakra-button css-zndnet"
          data-part="page-trigger"
          data-scope="pagination"
          data-testid="hds.minimal-pagination.page.control"
          id="pagination::r0::item:4"
          onClick={[Function]}
          type="button"
        >
          4
        </button>
        <button
          aria-label="page 5"
          className="chakra-button css-zndnet"
          data-part="page-trigger"
          data-scope="pagination"
          data-testid="hds.minimal-pagination.page.control"
          id="pagination::r0::item:5"
          onClick={[Function]}
          type="button"
        >
          5
        </button>
        <button
          className="chakra-button css-zndnet"
          data-part="ellipsis"
          data-scope="pagination"
          id="pagination::r0::ellipsis:5"
          type="button"
        >
          ...
        </button>
        <button
          aria-label="last page, page 20"
          className="chakra-button css-zndnet"
          data-part="page-trigger"
          data-scope="pagination"
          data-testid="hds.minimal-pagination.page.control"
          id="pagination::r0::item:20"
          onClick={[Function]}
          type="button"
        >
          20
        </button>
      </div>
      <button
        aria-label="Go to next page"
        className="chakra-button css-wklby5"
        data-part="next-page-trigger"
        data-scope="pagination"
        data-testid="hds.minimal-pagination.next.button"
        id="pagination::r0::next"
        onClick={[Function]}
        type="button"
      >
        Next
        <svg
          className="chakra-icon chakra-icon css-wuvk2m"
          focusable={false}
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4.16675 9.99996H15.8334M15.8334 9.99996L10.0001 4.16663M15.8334 9.99996L10.0001 15.8333"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="1.66667"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
